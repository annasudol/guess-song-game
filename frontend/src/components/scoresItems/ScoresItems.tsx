import { getToken } from '../../state/user/selectors';
import React, { FunctionComponent, ReactElement, useEffect, useState, useMemo } from 'react';
import { useSelector } from "react-redux";
import { useFetchPlaylist } from "../../hooks";
import { PlaylistItem, ScoresDBPropsGrouped, ErrorMessage } from ".."

export const ScoresItems: FunctionComponent<{ playlistIds: string[], scoresDB: ScoresDBPropsGrouped[] }> = ({ playlistIds, scoresDB }): ReactElement => {
    const token = useSelector(getToken);
    const [data, setData] = useState([]);
    const { playlistInfo, error } = useFetchPlaylist(playlistIds, token);

    useEffect(() => {
        !error && setData([...data, ...playlistInfo]);
    }, [playlistInfo, error]);

    if (error) {
        return (<ErrorMessage message="ups..., Error, something went wrong" />)
    }


    return (
        <div className="flex w-full flex-wrap">
            {!scoresDB.length && (
                <h3 className="font-alba text-orange-500 text-3xl mt-10">You don&apos;t have any games records yet</h3>
            )}
            {scoresDB?.map((info: ScoresDBPropsGrouped) => {
                const id = Object.keys(info)[0];
                const date = info[id].date;
                const options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
                const dateFormatted = new Date(date).toLocaleDateString(undefined, options);
                const playlist = data.filter(item => Object.keys(item)[0] === id)[0];
                const name = playlist && Object.values(playlist)[0]['name'];

                return (
                    <div className="w-full md:w-1/3 min-w-64 flex" key={id}>
                        <div className="w-1/3 max-w-24">
                            {playlist && <PlaylistItem playlist={playlist} />}
                        </div>
                        <div className="w-2/3 pl-4 pt-6">
                            <h3 className="font-alba text-orange-500 text-xl -mb-2">
                                {name?.length > 12 ? `${name?.slice(0, 12)}...` : name}
                            </h3>
                            <span className="font-OpenSans text-light-gray text-xs opacity-50">{dateFormatted}</span>
                            <p className="font-albaSuper text-jellyBean text-2xl">{info[id].points} p.</p>
                        </div>
                    </div>
                );

            })}
        </div>
    );
};