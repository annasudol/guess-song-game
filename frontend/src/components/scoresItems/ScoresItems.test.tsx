import { shallow } from 'enzyme';
import React from 'react';

import { ScoresItems } from './ScoresItems';
import toJson from 'enzyme-to-json';
jest.mock('react-redux');
const useSelector = require('react-redux').useSelector as jest.Mock;
const playlistIds = ['2GiFTnxorbfpgrXJYek2hv', '37i9dQZF1DXc3KygMa1OE7'];
const scoresDB = [{ '2GiFTnxorbfpgrXJYek2hv': { date: '2020-04-15T14:19:58.738Z', points: -30 } }, { '37i9dQZF1DXc3KygMa1OE7': { date: '2020-04-15T16:50:33.987Z', points: -40 } }];
describe('component', (): void => {
  describe('ScoresItems', (): void => {
    useSelector.mockReturnValue({
      getToken: 'BQAY_Ky0EiDo4z-U-oC2rYg5aF_kcSUhbukZ7BOJkXYQJzI-ChNz0hrULaIAoTmZ6s3_W_8n2LBh6_OKm5vgtvthrWLaPnGwYutjlGFW4cwQKqvN3ovaJmksGuzj5inFTH3UpEJ6_biTyDr_0w52EBirZQLGDN42mtS7ygfv2Nz',
    });
    it('match snapshot', (): void => {
      const wrapper = shallow(<ScoresItems playlistIds={playlistIds} scoresDB={scoresDB} />);

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
