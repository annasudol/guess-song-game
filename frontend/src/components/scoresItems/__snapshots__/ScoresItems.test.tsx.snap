// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component ScoresItems match snapshot 1`] = `
<div
  className="flex w-full flex-wrap"
>
  <div
    className="w-full md:w-1/3 min-w-64 flex"
    key="2GiFTnxorbfpgrXJYek2hv"
  >
    <div
      className="w-1/3 max-w-24"
    />
    <div
      className="w-2/3 pl-4 pt-6"
    >
      <h3
        className="font-alba text-orange-500 text-xl -mb-2"
      />
      <span
        className="font-OpenSans text-light-gray text-xs opacity-50"
      >
        Wednesday, 15 April 2020
      </span>
      <p
        className="font-albaSuper text-jellyBean text-2xl"
      >
        -30
         p.
      </p>
    </div>
  </div>
  <div
    className="w-full md:w-1/3 min-w-64 flex"
    key="37i9dQZF1DXc3KygMa1OE7"
  >
    <div
      className="w-1/3 max-w-24"
    />
    <div
      className="w-2/3 pl-4 pt-6"
    >
      <h3
        className="font-alba text-orange-500 text-xl -mb-2"
      />
      <span
        className="font-OpenSans text-light-gray text-xs opacity-50"
      >
        Wednesday, 15 April 2020
      </span>
      <p
        className="font-albaSuper text-jellyBean text-2xl"
      >
        -40
         p.
      </p>
    </div>
  </div>
</div>
`;
